rllib/agents/ppo/ppo.py

```diff
@@ -410,6 +418,86 @@ def get_default_policy_class(self, config: TrainerConfigDict) -> Type[Policy]:
         else:
             return PPOTFPolicy
 
+    @override(Trainer)
+    def training_iteration(self) -> ResultDict:
+        # Collects SampleBatches in parallel and synchronously
+        # from the Trainer's RolloutWorkers until we hit the
+        # configured `train_batch_size`.
+        sample_batches = []
+        num_env_steps = 0
+        num_agent_steps = 0
+        while (not self._by_agent_steps and num_env_steps < self.config[
+            "train_batch_size"]) or (
+            self._by_agent_steps and num_agent_steps < self.config["train_batch_size"]
+        ):
+            new_sample_batches = synchronous_parallel_sample(self.workers)
+            sample_batches.extend(new_sample_batches)
+            num_env_steps += sum(len(s) for s in new_sample_batches)
+            num_agent_steps += sum(s.agent_steps() for s in new_sample_batches)
+        self._counters[NUM_ENV_STEPS_SAMPLED] += num_env_steps
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += num_agent_steps
+
+        # Combine all batches at once and treat as multi-agent from here on.
+        train_batch = SampleBatch.concat_samples(sample_batches)
+        train_batch = train_batch.as_multi_agent()
```

Comment:
not really related, but it feels weird to me to have every agent do this as_multi_agent() call everywhere.

Reply:
I feel like for the entire ["collect batches until train_batch_size reached"]-problem, we should have a utility function. It all comes from the fact that we return different things from `sample()`, depending on whether the setup is multi-agent or not. I feel like we should always return MultiAgentBatch from `sample()`, never just SampleBatch.